### PR TITLE
Sane handling and prevention of default state leaking for gems in ruby_execute.

### PR DESCRIPTION
**Pull request by tylercloke**
Ping @chef-cookbooks/delivery-migration-team

### CODE REVIEW

**Comment by chef-delivery** [comment]
This PR has passed 'Verify' and is ready for review and approval!
  Use: '@delivery approve' when code review is complete.

**Comment by chef-delivery** [comment]
This PR has passed 'Verify' and is ready for review and approval!
  Use: '@delivery approve' when code review is complete.

**Comment by chef-delivery** [comment]
This PR has passed 'Verify' and is ready for review and approval!
  Use: '@delivery approve' when code review is complete.

File: libraries/ruby_execute.rb [hunk 1]
```diff
@@ -23,6 +23,15 @@
...
     resource_name :ruby_execute
+
+    # Specifies the GEM_HOME.
+    # This prevents global gem state by always setting a
+    # default not local to global ruby installation cache.
+    # We default this to #{CWD}/gem_cache if not specified by user.
+    # This will be where ruby looks for gems, and where bundle install will
+    # install by default. So it will "just work" if the user bundle installs
+    # and bundle.
+    attribute :gem_home, kind_of: String
```
Comment (schisamo) [inline libraries/ruby_execute.rb]: We should set the default in here. This should work:

```ruby
attribute :gem_home,
              kind_of: String,
              default: lazy { |r| ::File.join((r.cwd || Dir.pwd), 'gem_cache')  }
```

File: libraries/ruby_execute.rb [hunk 2]
```diff
@@ -56,6 +65,17 @@ def environment
       # ensure we don't destroy the `PATH` value set by the user
       existing_path = environment.delete('PATH')
       environment['PATH'] = [ruby_path, existing_path].compact.join(::File::PATH_SEPARATOR)
+
+      # defaults to #{CWD}/gem_cache if not defined by user
+      gem_home = if new_resource.gem_home.nil?
```
Comment (elliott-davis) [inline libraries/ruby_execute.rb]: :+1: to @schisamo's comment above since it would kill this block.

**Comment by chef-delivery** [comment]
This PR has passed 'Verify' and is ready for review and approval!
  Use: '@delivery approve' when code review is complete.

**Comment by schisamo** [comment]
@delivery approve

**Comment by chef-delivery** [comment]
Change: 20351da5-0893-4aae-a57d-efbcfffe99c7 approved by: @schisamo

**Comment by chef-delivery** [comment]
This PR has passed 'Acceptance' and is ready to be delivered!
  Use: '@delivery deliver' when validated in acceptance.

**Comment by schisamo** [comment]
@delivery deliver

**Comment by chef-delivery** [comment]
Change: 20351da5-0893-4aae-a57d-efbcfffe99c7 delivered by: @schisamo

**Pull request merged**